From 78933f90212129ac79860c7adca017d5dc7d9618 Mon Sep 17 00:00:00 2001
From: Eshwar Stalin <estalin@chromium.org>
Date: Fri, 16 Jan 2026 09:59:57 -0800
Subject: [PATCH] [Vertical Tabs] Fixing a crash during drag on windows with pinned tabs

This is a minor crash fix for drag and drop on vertical tab windows with
pinned tabs. As part of this, removed some of the interface functions in
TabDragContext which weren't needed or unused.

Change-Id: I9848f7410c1f29fdbeb8c9930a263a267057cc2e
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7488337
Reviewed-by: Kaan Alsan <alsan@chromium.org>
Commit-Queue: Eshwar Stalin <estalin@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1570472}
---

--- a/chrome/browser/ui/views/tabs/dragging/tab_drag_context.h
+++ b/chrome/browser/ui/views/tabs/dragging/tab_drag_context.h
@@ -17,7 +17,6 @@
 #include "ui/views/view.h"
 
 class Tab;
-class TabGroupHeader;
 class TabSlotView;
 class TabStrip;
 class TabStripModel;
@@ -130,10 +129,8 @@ class TabDragContext : public views::Vie
   virtual TabSlotView* GetTabForContents(content::WebContents* contents) = 0;
   virtual content::WebContents* GetContentsForTab(TabSlotView* view) = 0;
   virtual bool IsTabDetachable(const TabSlotView* view) const = 0;
-  virtual int GetTabCount() const = 0;
   virtual bool IsTabPinned(const TabSlotView* tab) const = 0;
-  virtual int GetPinnedTabCount() const = 0;
-  virtual TabGroupHeader* GetTabGroupHeader(
+  virtual TabSlotView* GetTabGroupHeader(
       const tab_groups::TabGroupId& group) const = 0;
   virtual TabStripModel* GetTabStripModel() = 0;
 
--- a/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.cc
+++ b/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.cc
@@ -1322,7 +1322,8 @@ void TabDragController::AttachToNewConte
   // Insert at any valid index in the tabstrip. We'll fix up the insertion
   // index in MoveAttached() later, if we're transitioning to kDraggingTabs;
   // if we're transitioning to kDraggingWindow this is the correct index, 0.
-  size_t index = attached_context_->GetPinnedTabCount();
+  size_t index =
+      attached_context_->GetTabStripModel()->IndexOfFirstNonPinnedTab();
 
   base::AutoReset<bool> setter(&is_mutating_, true);
 
@@ -1741,10 +1742,10 @@ std::vector<TabSlotView*> TabDragControl
         return {};
       }
 
-      TabGroupHeader* header =
+      TabSlotView* header_view =
           context->GetTabGroupHeader(tab_drag_datum.tab_group_data->group_id);
-      CHECK(header);
-      views.push_back(header);
+      CHECK(header_view);
+      views.push_back(header_view);
     }
   }
   return views;
--- a/chrome/browser/ui/views/tabs/dragging/test/mock_tab_drag_context.h
+++ b/chrome/browser/ui/views/tabs/dragging/test/mock_tab_drag_context.h
@@ -33,10 +33,8 @@ class MockTabDragContext : public TabDra
               IsTabDetachable,
               (const TabSlotView* view),
               (const, override));
-  MOCK_METHOD(int, GetTabCount, (), (const, override));
   MOCK_METHOD(bool, IsTabPinned, (const TabSlotView* tab), (const, override));
-  MOCK_METHOD(int, GetPinnedTabCount, (), (const, override));
-  MOCK_METHOD(TabGroupHeader*,
+  MOCK_METHOD(TabSlotView*,
               GetTabGroupHeader,
               (const tab_groups::TabGroupId& group),
               (const, override));
--- a/chrome/browser/ui/views/tabs/tab_strip.cc
+++ b/chrome/browser/ui/views/tabs/tab_strip.cc
@@ -403,6 +403,8 @@ class TabStrip::TabDragContextImpl : pub
     return tab_strip_->GetModelIndexOf(view);
   }
 
+  int GetTabCount() const { return tab_strip_->GetTabCount(); }
+
   // TabDragContext:
   TabDragContext* GetContextForNewBrowser(
       BrowserView* browser_view) const override {
@@ -433,17 +435,11 @@ class TabStrip::TabDragContextImpl : pub
              GetIndexOf(view) == 0);
   }
 
-  int GetTabCount() const override { return tab_strip_->GetTabCount(); }
-
   bool IsTabPinned(const TabSlotView* tab) const override {
     return tab_strip_->IsTabPinned(tab);
   }
 
-  int GetPinnedTabCount() const override {
-    return tab_strip_->NumPinnedTabsInModel();
-  }
-
-  TabGroupHeader* GetTabGroupHeader(
+  TabSlotView* GetTabGroupHeader(
       const tab_groups::TabGroupId& group) const override {
     return tab_strip_->group_header(group);
   }
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
@@ -280,16 +280,7 @@ bool VerticalTabDragHandlerImpl::IsTabDe
   return true;
 }
 
-int VerticalTabDragHandlerImpl::GetTabCount() const {
-  return dragged_tabs_.size();
-}
-
-int VerticalTabDragHandlerImpl::GetPinnedTabCount() const {
-  // TODO(crbug.com/439963720): Support dragging pinned tabs.
-  return 0;
-}
-
-TabGroupHeader* VerticalTabDragHandlerImpl::GetTabGroupHeader(
+TabSlotView* VerticalTabDragHandlerImpl::GetTabGroupHeader(
     const tab_groups::TabGroupId& group) const {
   // TODO(crbug.com/439963720): Support dragging tab groups.
   return nullptr;
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.h
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.h
@@ -88,9 +88,7 @@ class VerticalTabDragHandlerImpl : publi
   content::WebContents* GetContentsForTab(TabSlotView* tab) override;
   bool IsTabDetachable(const TabSlotView* view) const override;
   bool IsTabPinned(const TabSlotView* tab) const override;
-  int GetTabCount() const override;
-  int GetPinnedTabCount() const override;
-  TabGroupHeader* GetTabGroupHeader(
+  TabSlotView* GetTabGroupHeader(
       const tab_groups::TabGroupId& group) const override;
   TabStripModel* GetTabStripModel() override;
   TabDragController* GetDragController() override;
